File: shared/core/src/main/java/com/lacker/micros/domain/infrastructure/uuid/UUIDGenerator.java
package com.lacker.micros.domain.infrastructure.uuid;

import java.util.UUID;

public class UUIDGenerator {

    private static final UUIDGenerator INSTANCE = new UUIDGenerator();

    private static short counter = (short) 0;

    public static String sequentialUUIDString() {
        return INSTANCE.generateUUIDString();
    }

    private String generateUUIDString() {
        return generateUUID().toString().toUpperCase();
    }

    private UUID generateUUID() {
        long mostSignificantBits = (System.currentTimeMillis() << 20) | (getCount());
        long leastSignificantBits = UUID.randomUUID().getLeastSignificantBits();

        return new UUID(mostSignificantBits, leastSignificantBits);
    }

    private short getCount() {
        synchronized (UUIDGenerator.class) {
            if (counter < 0) {
                counter = 0;
            }
            return counter++;
        }
    }
}

File: library/core/src/main/kotlin/com/mlacker/micros/utils/id/UUIDGenerator.kt
package com.mlacker.micros.utils.id

import java.util.*
import java.util.concurrent.atomic.AtomicInteger

class UUIDGenerator : IdGenerator<String> {

    private val counter = AtomicInteger()

    override fun generate() = generateUUID().toString().toUpperCase()

    private fun generateUUID(): UUID {
        val mostSignificantBits = System.currentTimeMillis() shl 20 or count.toLong()
        val leastSignificantBits = UUID.randomUUID().leastSignificantBits
        return UUID(mostSignificantBits, leastSignificantBits)
    }

    private val count: Int
        get() = counter.getAndIncrement() and 0x0FFF

    companion object {
        private val INSTANCE = UUIDGenerator()

        fun generateId(): String {
            return INSTANCE.generate()
        }
    }
}